politeiad/backendv2/tstorebe/plugins/pi/cmds.go

```diff
@@ -0,0 +1,230 @@
+		Timestamp: time.Now().Unix(),
+	}
+	err = p.billingStatusSave(token, bsc)
+	if err != nil {
+		return "", nil
+	}
+
+	// Prepare reply
+	sbsr := pi.SetBillingStatusReply{
+		Timestamp: bsc.Timestamp,
+	}
+	reply, err := json.Marshal(sbsr)
+	if err != nil {
+		return "", err
+	}
+
+	return string(reply), nil
+}
+
+// billingStatusSave saves a BillingStatusChange to the backend.
+func (p *piPlugin) billingStatusSave(token []byte, bsc pi.BillingStatusChange) error {
+	// Prepare blob
+	be, err := convertBlobEntryFromBillingStatus(bsc)
+	if err != nil {
+		return err
+	}
+
+	// Save blob
+	return p.tstore.BlobSave(token, *be)
+}
+
+// billingStatuses returns all BillingStatusChange for a record.
+func (p *piPlugin) billingStatuses(token []byte) ([]pi.BillingStatusChange, error) {
+	// Retrieve blobs
+	blobs, err := p.tstore.BlobsByDataDesc(token,
+		[]string{dataDescriptorBillingStatus})
+	if err != nil {
+		return nil, err
+	}
+
+	// Decode blobs
+	statuses := make([]pi.BillingStatusChange, 0, len(blobs))
+	for _, v := range blobs {
+		a, err := convertBillingStatusFromBlobEntry(v)
+		if err != nil {
+			return nil, err
+		}
+		statuses = append(statuses, *a)
+	}
+
+	// Sanity check. They should already be sorted from oldest to
+	// newest.
+	sort.SliceStable(statuses, func(i, j int) bool {
+		return statuses[i].Timestamp < statuses[j].Timestamp
+	})
+
+	return statuses, nil
+}
+
+// tokenVerify verifies that a token that is part of a plugin command payload
+// is valid. This is applicable when a plugin command payload contains a
+// signature that includes the record token. The token included in payload must
+// be a valid, full length record token and it must match the token that was
+// passed into the politeiad API for this plugin command, i.e. the token for
+// the record that this plugin command is being executed on.
+func tokenVerify(cmdToken []byte, payloadToken string) error {
+	pt, err := tokenDecode(payloadToken)
+	if err != nil {
+		return backend.PluginError{
+			PluginID:     pi.PluginID,
+			ErrorCode:    uint32(pi.ErrorCodeTokenInvalid),
+			ErrorContext: util.TokenRegexp(),
+		}
+	}
+	if !bytes.Equal(cmdToken, pt) {
+		return backend.PluginError{
+			PluginID:  pi.PluginID,
+			ErrorCode: uint32(pi.ErrorCodeTokenInvalid),
+			ErrorContext: fmt.Sprintf("payload token does not "+
+				"match command token: got %x, want %x", pt,
+				cmdToken),
+		}
+	}
+	return nil
+}
+
+func convertSignatureError(err error) backend.PluginError {
+	var e util.SignatureError
+	var s pi.ErrorCodeT
+	if errors.As(err, &e) {
+		switch e.ErrorCode {
+		case util.ErrorStatusPublicKeyInvalid:
+			s = pi.ErrorCodePublicKeyInvalid
+		case util.ErrorStatusSignatureInvalid:
+			s = pi.ErrorCodeSignatureInvalid
+		}
+	}
+	return backend.PluginError{
+		PluginID:     pi.PluginID,
+		ErrorCode:    uint32(s),
+		ErrorContext: e.ErrorContext,
+	}
+}
+
+func convertBlobEntryFromBillingStatus(bsc pi.BillingStatusChange) (*store.BlobEntry, error) {
```

Comment:
I changed the naming convention for these in 1406.  I would change this function name to:
```
// billingStatusEncode encodes a BillingStatusChange into a BlobEntry.
func billingStatusEncode(bsc pi.BillingStatusChange) (*store.BlobEntry, error) {
```

And then the conversion function below to:
```
// billingStatusDecode decodes a BlobEntry into a BillingStatusChange.
func billingStatusDecode(be store.BlobEntry) (*pi.BillingStatusChange, error) {
```

Then you can list these up with the other `BillingStatus` functions.  It makes it easy to work with when all the billing status functions are listed right there together:

```
billingStatusEncode
billingStatusDecode
billingStatusSave
billingStatuses
```

Reply:
awesome stuff 💯